x-pack/plugins/endpoint/server/routes/resolver/utils/tree.ts

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { ResolverEvent, Node, NodeStats, NodePagination } from '../../../../common/types';
+import { extractEntityID, extractParentEntityID } from './normalize';
+import { buildPaginationCursor } from './pagination';
+
+type ExtractFunction = (event: ResolverEvent) => string | undefined;
+
+function createNode(id: string): Node {
+  return { id, children: [], pagination: {}, events: [], alerts: [], lifecycle: [] };
+}
+
+// This class aids in constructing a tree of process events. It works in the following way:
+//
+// 1. We construct a tree structure starting with the root node for the event we're requesting.
+// 2. We leverage the ability to pass hashes and arrays by reference to construct a fast cache of
```

Comment:
what does this mean?

Reply:
I think what this means is we're using a map of `entity_id` to `Node`, so that we don't have to traverse the tree when adding new information.